### PR TITLE
Add the new RPC function existsmempooltxs

### PR DESCRIPTION
**Pull request by cjepson**
A new RPC function has been added to the daemon to quickly check for
the existence of transactions in the mempool. It handles raw encoded
hex containing transaction hashes and outputs a raw encoded hex of
bit flags. This makes the output size as small as possible while
speeding up the JSON encoding/decoding of the data.

### CODE REVIEW

**Comment by jcvernaleo** [comment]
Need to fix the tests for this:
jcv@triforce dcrd $ go test
--- FAIL: TestHelp (0.00s)
        rpcserverhelp_test.go:18: RPC handler defined for method 'existsmempooltxs' without also specifying result types
        rpcserverhelp_test.go:45: Failed to generate help for method 'existsmempooltxs': no result types specified for method existsmempooltxs
FAIL
exit status 1
FAIL    github.com/decred/dcrd  0.029s

**Comment by jcvernaleo** [comment]
Other than the test issue, tACK.

**Comment by jcvernaleo** [comment]
Also needs a rebase now.

File: mempool.go [hunk 1]
```diff
@@ -925,6 +925,31 @@ func (mp *txMemPool) HaveTransaction(hash *chainhash.Hash) bool {
...
 
+// haveTransactions returns whether or not the passed transactions already exist
+// in the main pool or in the orphan pool.
+//
+// This function MUST be called with the mempool lock held (for reads).
+func (mp *txMemPool) haveTransactions(hashes []*chainhash.Hash) []bool {
+	have := make([]bool, len(hashes))
+	for i := range hashes {
+		have[i] = mp.isTransactionInPool(hashes[i]) ||
+			mp.isOrphanInPool(hashes[i])
```
Comment (dajohi) [inline mempool.go]: have[i] = mp.haveTransaction(hashes[i])

File: rpcserver.go [hunk 2]
```diff
@@ -1557,6 +1558,53 @@ func handleExistsLiveTickets(s *rpcServer, cmd interface{},
...
+// handleExistsMempoolTxs implements the existsmempooltxs command.
+func handleExistsMempoolTxs(s *rpcServer, cmd interface{},
+	closeChan <-chan struct{}) (interface{}, error) {
+	c := cmd.(*dcrjson.ExistsMempoolTxsCmd)
+
+	txHashBlob, err := hex.DecodeString(c.TxHashBlob)
+	if err != nil {
+		return nil, &dcrjson.RPCError{
+			Code:    dcrjson.ErrRPCDecodeHexString,
+			Message: "bad transaction hash blob (unparseable)",
```
Comment (dajohi) [inline rpcserver.go]: Message: fmt.Sprintf("bad transaction hash blob: %v", err),

File: rpcserver.go [hunk 3]
```diff
@@ -1557,6 +1558,53 @@ func handleExistsLiveTickets(s *rpcServer, cmd interface{},
...
+		}
+	}
+
+	// It needs to be an exact number of hashes.
+	if len(txHashBlob)%32 != 0 {
+		return nil, &dcrjson.RPCError{
+			Code:    dcrjson.ErrRPCDecodeHexString,
+			Message: "bad transaction hash blob (bad length)",
+		}
```
Comment (dajohi) [inline rpcserver.go]: Message: fmt.Sprintf("bad transaction hash blob: %d != 0", len(txHashBlock)%32),

File: rpcserver.go [hunk 4]
```diff
@@ -1557,6 +1558,53 @@ func handleExistsLiveTickets(s *rpcServer, cmd interface{},
...
+	for i := 0; i < hashesLen; i++ {
+		hashes[i], err = chainhash.NewHash(
+			txHashBlob[i*chainhash.HashSize : (i+1)*chainhash.HashSize])
+	}
+
+	exists := s.server.txMemPool.HaveTransactions(hashes)
+	if len(exists) != hashesLen {
+		return nil, &dcrjson.RPCError{
+			Code:    dcrjson.ErrRPCDatabase,
+			Message: "output of ExistsMempoolTxs wrong size",
```
Comment (dajohi) [inline rpcserver.go]: Message: fmt.Sprintf("output of ExistsMempoolTxs wrong size: %d != %d", len(exists), hashesLen),

File: rpcserver.go [hunk 5]
```diff
@@ -1557,6 +1558,53 @@ func handleExistsLiveTickets(s *rpcServer, cmd interface{},
...
+			Code:    dcrjson.ErrRPCDecodeHexString,
+			Message: "bad transaction hash blob (bad length)",
+		}
+	}
+
+	hashesLen := len(txHashBlob) / 32
+	hashes := make([]*chainhash.Hash, hashesLen)
+	for i := 0; i < hashesLen; i++ {
+		hashes[i], err = chainhash.NewHash(
+			txHashBlob[i*chainhash.HashSize : (i+1)*chainhash.HashSize])
```
Comment (dajohi) [inline rpcserver.go]: unhandled error

File: rpcserver.go [hunk 6]
```diff
@@ -1524,7 +1525,7 @@ func handleExistsLiveTickets(s *rpcServer, cmd interface{},
 	if len(txHashBlob)%32 != 0 {
 		return nil, &dcrjson.RPCError{
 			Code:    dcrjson.ErrRPCDecodeHexString,
-			Message: "bad transaction hash blob (bad length)",
+			Message: "bad ticket hash blob (bad length)",
```
Comment (dajohi) [inline rpcserver.go]: Message: fmt.Sprintf("bad ticket hash blob (bad length): %d != 0", len(txHashBlob)%32),

**Comment by dajohi** [comment]
OK for squash and merge

**Pull request merged**